src/app/recipes/recipe-edit/recipe-edit.component.ts

```diff
@@ -1,8 +1,93 @@
-import { Component } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
+import { FormArray, FormControl, FormGroup, Validators } from '@angular/forms';
+import { UnitsService } from '../../shared/units.service';
+import { FormToRecipeService } from '../form-to-recipe.service';
+import { RecipesService } from '../recipes.service';
+import { ActivatedRoute, Router } from '@angular/router';
+import { Recipe } from '../recipe.interface';
+import { RecipeFormService } from './recipe-form.service';
 
 @Component({
   selector: 'app-recipe-edit',
   templateUrl: './recipe-edit.component.html',
   styleUrl: './recipe-edit.component.scss',
 })
-export class RecipeEditComponent {}
+export class RecipeEditComponent implements OnInit {
+  recipeForm: FormGroup;
+  unitsList = this.units.units;
+  imagePreview: string;
+  id: number | null = this.route.snapshot.queryParams['id']
+    ? +this.route.snapshot.queryParams['id']
+    : null;
+  loadedRecipe: Recipe | null;
+  constructor(
+    private units: UnitsService,
+    private formToRecipe: FormToRecipeService,
+    private recipesService: RecipesService,
+    private route: ActivatedRoute,
+    private router: Router,
+    private form: RecipeFormService,
+  ) {}
+
+  ngOnInit(): void {
+    this.recipeForm = this.form.createEmptyForm();
+    if (this.id) {
+      this.loadedRecipe = this.recipesService.recipeById(this.id);
+      this.form.recreateRecipeForm(this.recipeForm, this.loadedRecipe);
+    }
+
+    if (
+      this.loadedRecipe &&
+      this.loadedRecipe.imgs &&
+      typeof this.loadedRecipe.imgs[0] == 'string'
+    ) {
+      this.imagePreview = this.loadedRecipe.imgs[0];
+    }
+  }
```

Comment:
Extract contents of this hook into separate methods, keep ngoninit clean always. Also, when using simple if it's better to escape the function when the condition is not met rather than putting code within the block if possible e.g

instead
```
if(conditition) { do(); }
```

do

```
if(!conditition) { 
  return;
 }

do();
```